examples/snake/main.go

```diff
@@ -0,0 +1,206 @@
+// Copyright 2020 The Ebiten Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// +build example jsgo
+
+package main
+
+import (
+	"fmt"
+	"image/color"
+	"log"
+	"math/rand"
+
+	"github.com/hajimehoshi/ebiten"
+	"github.com/hajimehoshi/ebiten/ebitenutil"
+	"github.com/hajimehoshi/ebiten/inpututil"
+)
+
+const (
+	screenWidth  = 640
+	screenHeight = 480
+	nx           = screenWidth / 10
+	ny           = screenHeight / 10
+)
+
+// SnakePositionOnGrid ...
+type SnakePositionOnGrid struct {
+	X int64
+	Y int64
+}
+
+// Game ...
+type Game struct {
+	moveDirection int8
+	appleX        int64
+	appleY        int64
+	snakeBody     []SnakePositionOnGrid
+	timer         uint64
+	speed         int8
+	score         int64
+	bestScore     int64
+	level         int8
+}
+
+func (g *Game) collidesWithApple() bool {
+	return g.snakeBody[0].X == g.appleX &&
+		g.snakeBody[0].Y == g.appleY
+}
+
+func (g *Game) collidesWithSelf() bool {
+	for _, v := range g.snakeBody[1:] {
+		if g.snakeBody[0].X == v.X &&
+			g.snakeBody[0].Y == v.Y {
+			return true
+		}
+	}
+	return false
+}
+
+func (g *Game) collidesWithWall() bool {
+	return g.snakeBody[0].X < 0 ||
+		g.snakeBody[0].Y < 0 ||
+		g.snakeBody[0].X >= nx ||
+		g.snakeBody[0].Y >= ny
+}
+
+func (g *Game) updateTimer() {
+	g.timer++
+}
+
+func (g *Game) needsToMoveSnake() bool {
+	return g.timer%uint64(g.speed) == 0
+}
+
+func (g *Game) reset() {
+	g.appleX = 30
+	g.appleY = 30
+	g.speed = 4 // means 83.33 ms
+	g.snakeBody = g.snakeBody[:1]
+	g.snakeBody[0].X = nx / 2
+	g.snakeBody[0].Y = ny / 2
+	g.score = 0
+	g.level = 1
+	g.moveDirection = 0
+}
+
+// Update ...
+func (g *Game) Update(screen *ebiten.Image) error {
+
```

Comment:
Don't use 1,2,3,4 as a direction. Let's define constants in this function use them like this:

```
const (
   dirLeft = iota
   dirRight
   dirDown
   dirUp
)
```